text/0000-task-builder.md

```diff
@@ -0,0 +1,200 @@
+
+# Motivation
+[motivation]: #motivation
+
+The current `Task` trait passes `self` as a shared reference (`&self`) to `fn perform`. Multiple users (https://github.com/neon-bindings/neon/pull/283) (https://github.com/neon-bindings/neon/pull/520) have found this ownership limiting. Many usages require a mutable or owned `self`. Work-arounds require error prone runtime checks (e.g., `Cell` and `Option`).
+
+While unlikely to cause issues, changing `&self` to `&mut self` is a breaking change. Users could have manually called `fn perform` in a context that relied on shared references.
+
+Instead of introducing a breaking change, a new API is introduced with an _owned_ `self`. The following addtional changes are made:
+
+* The new API relies on closures instead of a `trait`; thanks to many improvements in the borrow checker and closure ergonomics
+* The new API returns a single `Output` generic rather than separate `Output` and `Error` types. This simplifies infallible tasks.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Context` trait provides a `.task()` method for performing asynchronous tasks on a background thread and calling back to javascript on completion.
+
+The callback follows standard node.js callback style with an error as the first argument and the resolved value as the second argument.
+
+The following example introduces a method named `add_one` which accepts a `Number` and a callback, performing the operation asynchronously.
+
+```rust
+pub fn add_one(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let n = cx.argument::<JsNumber>(0)?.value();
+    let cb = cx.argument::<JsFunction>(1)?;
+
+    cx.task(move || {
+            let result = n + 1.0;
+
+            move |mut cx| Ok(cx.number(result))
+        })
+        .schedule(cb)
+}
+
+register_module!(mut cx, {
+    cx.export_function("add_one", addOne)
+});
+```
+
+The `add_one` method can be called from Javascript:
+
+```js
+const { addOne } from './native';
+
+addOne(41, (err, val) => {
+    if (err) {
+        console.error(err);
+    } else {
+        console.log(`Result: ${val}`);
+    }
+});
+```
+
+The `cx.task(...)` method creates a new `TaskBuilder`. It expects a single argument. This argument is a `closure` that will be executed on the `libuv` threadpool.
+
+Any setup that requires the V8 VM (e.g., getting an `f64` from a `Number`) must be performed outside of this closure.
+
+The closure is expected to return _another closure_. This higher-order method will be familiar to many Javascript developers. The returned closure is the `complete` method and will be performed back on the main Javascript thread. In the `complete` closure, the user may convert Rust types back to Javascript types.
+
+Finally, a `schedule` method is provided to submit the task to the queue. For conveinance, it returns `JsUndefined`, matching Javascript semantics for most asynchronous methods.
```

Comment:
I don't think I understand this. The `schedule` method returns `Handle<JsUndefined>`? Or the JS callback is wrapped to discard its result and return undefined? Or something else? It might be helpful to include the type signature.

Reply:
I'll write more examples. `schedule` returns `undefined` to remove a little more boilerplate of creating that value since that's the most common pattern.

Reply:
```rust
pub fn perform_closure_task(mut cx: FunctionContext) -> JsResult<JsValue> {
    let n = cx.argument::<JsNumber>(0)?.value();
    let cb = cx.argument::<JsFunction>(1)?;

    cx.task(move || {
            let result = n + 1.0;

            move |mut cx| Ok(cx.number(result))
        })
        .schedule(cb)
}
```

Reply:
I need to update the RFC, but, I'll made it return a `JsUndefined` upcast to `JsValue`. This makes it fit a few more situations. For example, class methods always return `JsValue`.